src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -21,49 +19,60 @@
 using Microsoft.CodeAnalysis.Shared.Utilities;
 using Microsoft.CodeAnalysis.SymbolSearch;
 using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Utilities;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.AddMissingImports
 {
     internal abstract class AbstractAddMissingImportsFeatureService : IAddMissingImportsFeatureService
     {
         protected abstract ImmutableArray<string> FixableDiagnosticIds { get; }
 
-        public async Task<bool> HasMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        /// <inheritdoc/>
+        public async Task<Project?> AddMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
         {
-            // Get the diagnostics that indicate a missing import.
-            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
-            if (diagnostics.IsEmpty)
+            var analysisResult = await AnalyzeAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return await AddMissingImportsAsync(analysisResult, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc/>
+        public async Task<Project?> AddMissingImportsAsync(AddMissingImportsAnalysisResult analysisResult, CancellationToken cancellationToken)
+        {
+            if (analysisResult.CanAddMissingImports)
             {
-                return false;
+                // Apply those fixes to the document.
+                var newDocument = await ApplyFixesAsync(analysisResult.Document, analysisResult.AddImportFixData, cancellationToken).ConfigureAwait(false);
```

Comment:
can pasting cause us to add things like references/nuget-packages?  if so... this can be extremely costly and not what users want.

Iwould honestly have two options we so we can limit this.  i.e. first option is "Add missing imports on paste" with a sub option of "Including missing references and Nuget packages".

The first can be on by default, i would make the second off by default.

Reply:
NuGet packages are not added as part of this service.